### PR TITLE
java formatter based on jdt

### PR DESCRIPTION
**Pull request by manuel-hegner**
I wanted to implement a java formatter based on the powerful formatter of eclipse. There is already a maven project that makes it easy to use with command line arguments at https://code.revelc.net/formatter-maven-plugin .

Sadly I was unable to follow the instructions well enough to actually test this. Hope somebody can help.

### CODE REVIEW

**Comment by pbrisbin** [comment]
> Sadly I was unable to follow the instructions well enough to actually test this

I would love to know the trouble you ran into, this looks pretty good! I think you just need to:

1. Build the Docker image:

   ```
   make jdt/Dockerfile.built
   ```

1. Run the test you added

   ```
   cram tests/jdt.t
   ```

   (Which will fail, but we can go from there!)

**Comment by manuel-hegner** [comment]
I built the docker image with
```bash
docker build --tag restyled/restyler-jdt jdt
```

but the test fails with
```
$ cram3 test/jdt.t
!
--- test/jdt.t
+++ test/jdt.t.err
@@ -1,6 +1,9 @@
   $ source "$TESTDIR/helper.sh"
-  If you don't see this, setup failed
+  /bin/sh: 2: source: not found
+  [127]

 jdt

   $ run_restyler jdt ./Foo.java
+  /bin/sh: 4: run_restyler: not found
+  [127]

# Ran 1 tests, 0 skipped, 1 failed.
```

**Comment by pbrisbin** [comment]
```
/bin/sh: 2: source: not found
```

Oh, it looks like you're tests are running in `sh` when we expect `bash`. And `source` vs `.` is one difference there. Do you have bash installed? If not, please install it, then try:

```
cram3 -i --shell=$(which bash) test/jdt.t
```

I'll update the docs about this.

**Comment by manuel-hegner** [comment]
Okay, I got it running so far. The problem now is that the `formatter-maven-plugin` expects to be run on a maven project in the root directory. Is there a way to specify this behaviour in the info.yaml?

**Comment by pbrisbin** [comment]
> The problem now is that the formatter-maven-plugin expects to be run on a maven project in the root directory.

I don't 100% understand this, but if you mean there are some project related metadata that is present in a real clone of a typical project, then I think that means this will work when run for real in a proper clone, but won't work in the test environment which only contains the fixture files.

If that's true, I think we need to tweak this test to create any of those files. Can you tell me what project files are needed and if you think they could be easily constructed by hand in the context of a test?

For context, cram tests just run shell commands, so (within reason) you should be able to do anything you need to to set up a proper test environment.

As a made-up example:

```cram
  $ source "$TESTDIR/helper.sh"
  If you don't see this, setup failed

jdt

  $ cat "{ is-a-maven-project: true }" > .mavenrc
  > run_restyler jdt ./Foo.java
  
```

I'd like to pursue getting the test running a little further, but it's not an "at all costs" requirement here. I'm happy to get this image pushed up so you can try it out for real, even if we can't get a good test over it at this time.

If you're interested, it's possible for you to push your image to your own Docker Hub and give it a try for real: https://github.com/restyled-io/restyled.io/wiki/Restyler-Versions#use-your-own-images

**Comment by manuel-hegner** [comment]
I would also like to get the test running. I added a minimalistic maven project to the fixtures. It consists of a `pom.xml` and a `src/main/java/Foo.java`.

I now get the problem that the docker formatter can't see the pom.xml and I am not sure why.
```
+  [ERROR] Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.10.0:format (default-cli): Goal requires a project to execute but there is no POM in this directory (/code). Please verify you invoked Maven from the correct directory. -> [Help 1]
+  /mnt/c/Users/manuel/Documents/hobby/restylers/test/../build/restyler-meta:29:in `run': ["docker", "run", "--rm", "--net", "none", "--volume", "/tmp/cramtests-d9ve5kf3/jdt.t/repo:/code", "restyled/restyler-jdt", "mvn", "net.revelc.code.formatter:formatter-maven-plugin:2.10.0:format"] failed (RuntimeError)
```

**Comment by pbrisbin** [comment]
Hmmm... I tried myself and was able to get this far:

```
+  [ERROR] Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.10.0:format (default-cli) on project fooArtifact: The parameters 'encoding' for goal net.revelc.code.formatter:formatter-maven-plugin:2.10.0:format are missing or invalid -> [Help 1]
```

I also notice you took the paths out of the execution, is there a way to specify the paths? Restyled requires that since we only re-format changed files.

**Comment by manuel-hegner** [comment]
Ah, I got it to run too.

The next problem is that `formatter-maven-plugin` expects the files in the format `-Dformatter.includes=Foo.java,Bar.java`, but it seems restyle always adds a space in front of the paths. I probably have to write a script to translate a call in the restyled style to a `formatter-maven-plugin` call.

**Comment by manuel-hegner** [comment]
Okay, so @martintaraz and I created a very simple script to translate the call. At least locally the test is green now. Right now the script is very basic though. The main problem is that it always expects the pom.xml in root and it looks for all files in `src/main/java`. Both things need to be improved before I would consider this PR ready for use.

How do we test with multiple files? We tested `run_restyler jdt ./Foo.java,./Bar.java` but that didn't seem to work. How does restyler call the formatter when there are multiple files?

**Comment by pbrisbin** [comment]
> How do we test with multiple files? We tested run_restyler jdt ./Foo.java,./Bar.java but that didn't seem to work. How does restyler call the formatter when there are multiple files?

Multiple files can be passed as separate arguments:

```
run_restyler jdt ./Foo.java ./Bar.java
```

**Comment by manuel-hegner** [comment]
I think this is ready for a review. It could still be expanded a bit but should support most maven cases now.

File: jdt/Formatter.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import java.io.File;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Formatter {
+
+	private static final List<String> PREFIXES = List.of(
```
Comment (pbrisbin) [inline jdt/Formatter.java]: These are needed, even if the full (repo-relative) paths are given in the command?

```
run_restyler jdt ./src/main/java/ShouldFormat1.java ./src/main/java/ShouldFormat2.java
```

Is it possible/common to use other prefixes in Java projects, and would that just get ignored by this Restyler?

**Comment by manuel-hegner** [comment]
Okay, there actually is a way to work from the project root directory. It should be much more portable now.

**Comment by manuel-hegner** [comment]
This should be ready to be merged now.

**Comment by pbrisbin** [comment]
Awesome! I'm going to make a new branch from it and clean up history / get it passing.

**Comment by pbrisbin** [comment]
Couple of quick questions:

- Is "JDT" the right name, or would "maven-formatter" be better?
- I see `include` has `pom.xml`, is the formatter meant to fix formatting in this file?
- I see `metadata.languages` includes HTML/CSS/JSON/XML -- does this formatter really work on these? If so, do you want to add `*.html`, etc to `include`?

**Comment by manuel-hegner** [comment]
- maven-formatter is just the name of the small projects that makes the JDT formatter easier to use. But maybe "Eclipse JDT" would be a better, since it is better known.
- the `pom.xml` in `include is wrong
- yes, it supports all of them and they should be also included

Should I change these things here or do you do this in #35 ?

**Comment by pbrisbin** [comment]
Thanks for the answers. I'll keep the name `jdt`, and I'll handle fixing `include`.

**Pull request closed without merging**